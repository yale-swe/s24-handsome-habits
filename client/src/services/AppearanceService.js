import { getQualityPoints } from "./PointsService";
import { emotions, tops, bottoms } from "./constants/resources";


/**
 * Returns the path to the image of the user's current emotion.
 * @returns {Any} The path to the image of the user's current emotion.
 */
export function getEmotionPath() {

    const emotion = getEmotion();

    if (emotion < 33.3) {
        return emotions["sad_face"];
    }
    else if (emotion < 66.6) {
        return emotions["neutral_face"];
    }
    else {
        return emotions["happy_face"];
    }
}

/**
 * Returns the user's current emotion value
 * @returns {Number} The user's current emotion value.
 */
export function getEmotion() {

    const userPoints = getQualityPoints();
    const emotion = getQualityPoints(userPoints).emotion_value;
    return emotion;

}

/**
 * Returns the path to the image of the user's current clothes.
 * @returns {Any} The path to the image of the user's current clothes.
 */
export function getClothesPath() {

    const clothes = getClothes();
    const clothesPath = {
        top: tops[clothes.top],
        bottom: bottoms[clothes.bottom],
        accessories: null
    };
    return clothesPath;


}

/**
 * Returns the user's current clothes.
 * @returns {JSON} The user's current clothes.
 */
export function getClothes() {

    return {
        top: "white_tshirt",
        bottom: "white_pants",
        accessories: null,
    }


}
