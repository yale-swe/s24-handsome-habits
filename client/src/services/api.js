import Axios from 'axios';

export default Axios.create({
    baseURL: process.env.EXPO_PUBLIC_SERVER_URL
});