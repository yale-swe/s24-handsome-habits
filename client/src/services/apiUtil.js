import Axios from 'axios';

const localhost = process.env.EXPO_PUBLIC_LOCAL_SERVER_URL || "http://localhost:8000";
export const serverURL = process.env.EXPO_PUBLIC_PROD == 1 ? process.env.EXPO_PUBLIC_PROD_SERVER_URL : localhost;

console.log("Server URL: ", serverURL);

export default Axios.create({
    baseURL: serverURL + "/api",
});

export const AuthApi = Axios.create({
    baseURL: serverURL + "/auth",
});