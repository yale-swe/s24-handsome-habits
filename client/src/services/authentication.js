import Api from "./api";
import AsyncStorage from "@react-native-async-storage/async-storage";

export default async function Authentication(response) {
  var user_data = null;
  const user = await AsyncStorage.getItem("@user");
  if (user) {
    logout();
    user_data = JSON.parse(user);
  }

  if (response?.type === "success") {
    console;
    user_data = await getUserInfoWithGoogle(
      response.authentication.accessToken
    );
  }
  console.log("User signed in:", user_data);
  return user_data;

  async function getUserInfoWithGoogle(token) {
    try {
      const user = await Api.post("/auth/login", { token: token });
      await AsyncStorage.setItem("@user", JSON.stringify(user.data));
      return JSON.stringify(user.data);
    } catch (e) {
      console.error("Error getting user info with Google", e);
    }
  }
}

export async function logout() {
  await AsyncStorage.removeItem("@user");
}
