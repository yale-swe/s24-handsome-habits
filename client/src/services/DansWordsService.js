import { getPointInfo } from "./PointsService";
import { getEmotion } from "./AppearanceService";

export const expressions = {

    Eating: [
        "Where can a dog find some good kibble around here?",
        "Did you hear that? I think my tummy just rumbled.",
        "It's been a ruff day... I could use a treat.",
        "I could go for a bone right about now...",
        "I'm peckish and parched!",
        "Feed me... please...",
    ],
    Sleeping: [
        "I'm dog tired today.",
        "I could use a nap right about now...",
        "Oh, what I'd do to nap in my dog house right now!",
        "I could really go for a cat-nap... or a dog-nap!",
        "I wish I could lay down for a bit...",
        "Was that a yawn just now?",
        "Oh boy, I'm exhausted...",
    ],
    Studying: [
        "I'm trying to expand my vocab beyond \"woof\" and \"bark.\"",
        "I wish I could learn something new today...",
        "My last report card didn't look so good...",
        "Wanna have a study sesh at my dog house?",
        "I saw the neatest book the other day!",
        "I'm feeling awfully studious today.",
        "Want to study together?",
    ],
    Exercising: [
        "I'm feeling a bit restless today...",
        "If we went swimming, I could teach you to doggy-paddle.",
        "I've got the zoomies! I would love to go run around!",
        "I would love the chance to stretch my paws!",
        "I could go for some belly rubs right now...",
        "Can we go to the park?",
        "Can we go for a walk?",
    ],
    Happy: [
        "Barking contest! Whoever barks louder wins!",
        "Woof! What a beautiful day it is!",
        "I'm feeling mighty handsome today.",
        "I'm having a dog-gone good day!",
        "I'm loving my cool threads!",
        "You deserve a round of a-paws!",
        "Does this shirt give me rizz?",
        "I want a new outfit!",
        "How's it going!",
        "Boola boola!",
    ]
}

/**
 * Gets a random expression based on emotion/lowest habit.
 * @returns {String} - The expression.
 */
export async function getExpression() {
    const points_values = await getEmotion();
    const emotion = points_values.wellness_points;
    if (emotion > 60) {
        const randomIndex = Math.floor(Math.random() * expressions["Happy"].length);
        const expression = expressions["Happy"][randomIndex];
        return expression;
    }
    else {
        const lowestHabit = await getLowestHabit();
        console.log(lowestHabit);
        const randomIndex = Math.floor(Math.random() * expressions[lowestHabit].length);
        const expression = expressions[lowestHabit][randomIndex];
        return expression;
    }

}

/**
 * Gets the lowest habit by percentage.
 * @returns {JSON} - The lowest habit.
 */
export async function getLowestHabit() {

    const userPoints = await getPointInfo();
    const eating = userPoints.eating_points / 25;
    const sleeping = userPoints.sleeping_points / 27;
    const studying = userPoints.studying_points / 22;
    const exercising = userPoints.exercise_points / 26;

    const lowest = Math.min(eating, sleeping, studying, exercising);

    if (lowest === eating) {
        return "Eating";
    } else if (lowest === sleeping) {
        return "Sleeping";
    } else if (lowest === studying) {
        return "Studying";
    } else {
        return "Exercising";
    }
}
