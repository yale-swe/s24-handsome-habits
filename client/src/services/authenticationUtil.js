import { AuthApi } from "./apiUtil";
import axios from "axios";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { StatusCodes } from "http-status-codes";

export default async function Authentication(response) {
  var user_data = null;
  const user = await AsyncStorage.getItem("@user");
  if (user) {
    logout();
    user_data = JSON.parse(user);
  }
  if (response?.type === "success") {
    user_data = await getUserInfoWithGoogle(
      response.authentication.accessToken,
    );

    console.log("User signed in:", user_data);
    return user_data ? user_data : null;
  }
  console.log("User login in failed");
  return null;
}

async function getUserInfoWithGoogle(token) {
  try {
    const user = await AuthApi.post("/google/login", { token: token });
    await AsyncStorage.setItem("user", JSON.stringify(user.data));
    return JSON.stringify(user.data);
  } catch (e) {
    console.error("Error getting user info with Google", e);
  }
}

export async function logout() {
  console.log("Logging out and deleting user cookies");
  // remove the cookie header from axios
  axios.defaults.headers.Cookie = "";
  await AsyncStorage.clear();
}

export async function CASLogout() {
  logout();
  await AuthApi.get("/cas/logout");
}

/**
 * @returns {} - The user data if found; otherwise, null.
 */
export async function LoginWithActiveSession() {
  let response = null;
  await AsyncStorage.removeItem("cookies");

  try {
    response = await AuthApi.get("/login", { withCredentials: true });
  } catch (err) {
    // Clear session if session is invalid / unauthorized user
    console.error(
      "Error logging in with stored cookies. Logging out",
      err.response.status,
    );
    if (err.response.status == StatusCodes.UNAUTHORIZED) {
      logout();
    }
  } finally {
    // eslint-disable-next-line no-unsafe-finally
    return response?.status === 200 ? response : null;
  }
}


/**
 * Logs in using a dummy account.
 * This function directly accesses the dummy login route and sets the dummy user data.
 * 
 * @returns {Object|null} - The user data if the login is successful; otherwise, null.
 */
export async function dummyLogin() {

  try{
    AsyncStorage.clear();
    console.log("Logging in with dummy user");
    await AsyncStorage.removeItem("cookies");
    const response = await AuthApi.post("/dummy_login");
    
    await AsyncStorage.setItem("@user", JSON.stringify(response.data.user));
    console.log("Logged in successfully with dummy user:", response.data.user);
    return response.data.user;

  }
  catch(err){
    console.error("Error logging in with dummy user", err);
    // return null;
  }


};