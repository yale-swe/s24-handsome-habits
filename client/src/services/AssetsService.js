import Api from "./apiUtil";
import { logout } from "./authenticationUtil";
import { StatusCodes } from "http-status-codes";
import AsyncStorage from "@react-native-async-storage/async-storage";

/**
 * @returns the user's assets JSON if found
 */
export async function getAssets() {
  try {
    const response = await Api.get("/assets");
    let assets = response.data.assets;

    // return retrieved assets and save them in AsyncStorage if they exist
    // otherwise assets stored in local storage
    if (assets) {
      AsyncStorage.setItem("assets", JSON.stringify(assets));
    } else {
      console.log(
        "No assets found: Trying to use locally saved assets in Asyc Storage",
      );
      assets = await JSON.parse(AsyncStorage.getItem("assets"));
    }

    return assets;
  } catch (err) {
    if (err.response?.status == StatusCodes.UNAUTHORIZED) {
      console.log("Error: Session is expired/invalid, so logout");
      logout();
      return null;
    } else {
      console.log(
        "Error finding assets. Returning locally saved assets in AsyncStorage",
      );
      return await JSON.parse(await AsyncStorage.getItem("assets"));
    }
  }
}

/**
 * Adds an asset to the user's assets
 * Function DOES NOT check whether the user owns the item
 * @param {String} assetType - One of [tops, bottoms, accessories] - Yes, plural, not singular
 * @param {String} assetName - the name of the asset
 * @returns the updated assets JSON if successful, null otherwise
 */
export async function addAsset(assetType, assetName) {
  try {
    const response = await Api.post("/assets/add", {
      assets: {
        type: assetType,
        name: assetName,
      },
    });

    const updatedAssets = response.data.assets;
    AsyncStorage.setItem("assets", JSON.stringify(updatedAssets));
    return updatedAssets;
  } catch (err) {
    if (err.response?.status == StatusCodes.UNAUTHORIZED) {
      logout();
    } else if (err.response?.status == StatusCodes.INTERNAL_SERVER_ERROR) {
      console.log(
        "Error adding asset; Returning locally saved assets in AsyncStorage",
      );
      return JSON.parse(await AsyncStorage.getItem("assets"));
    }
    return null;
  }
}

/**
 * Sets the active assets for the user
 * @param {JSON} newActiveAssets - Eg. {tops: "yale_shirt", bottoms: "yale_pants", accessories: "yale_hat"} - Yes, plural, not singular
 * @returns the updated assets JSON if successful, null otherwise
 */
export async function setActiveAssets(newActiveAssets) {
  try {
    const response = await Api.post("/assets/setActive", {
      assets: newActiveAssets,
    });

    const updatedAssets = response.data.assets; // entire assets object
    AsyncStorage.setItem("assets", JSON.stringify(updatedAssets));
    return updatedAssets;
  } catch (err) {
    // return previous assets in locally saved assets in AsyncStorage if unable to set new active assets
    if (err.response.status == StatusCodes.INTERNAL_SERVER_ERROR) {
      console.log("Error setting active asset.");
      console.log(
        "Assets not updated, so returning locally saved assets in AsyncStorage",
      );
      return JSON.parse(AsyncStorage.getItem("assets"));
    }

    if (err.response.status == StatusCodes.UNAUTHORIZED) {
      logout();
    } // logout if session is invalid

    return null;
  }
}
