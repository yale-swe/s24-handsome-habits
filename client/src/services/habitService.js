import Api from "./apiUtil";
import { logout } from "./authenticationUtil";
import { StatusCodes } from "http-status-codes";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { updatePointswithChange } from "./PointsService";
import { updateLastActivity } from "../services/LastActivityService"; // Adjust the path as necessary

/**
 * Adds a new habit for the user.
 * @param {JSON} newHabit - The new habit to add.
 * newHabit should contain:
 * {
 *   title: String,
 *   category_name: String,
 *   description: String,
 *   date_and_time: Date,
 *   details: {
 *      eating: { eating_tag: String, healthy_meal: Boolean },
        sleep: {sleep_duration: Number, quality_of_sleep: Number },
        study: { study_duration: Number, study_productivity: Number },
        workout: { workout_tag: "String, workout_duration: Number, workout_intensity: String }
 *   },
 *   date_and_time: Date,
 * }
 * @returns {JSON|null} - The added habit data if successful; otherwise, null.
 */
export async function addHabit(newHabit) {
    try {
        const habit_response = await Api.post("/habits/add", {
            habit: newHabit,
        }); // Post request to add a new exercise habit
        // Save new habit in client's local storage
        AsyncStorage.setItem("habit", JSON.stringify(habit_response.data));

        // Update last activity after successfully adding the habit
        const lastActivityUpdate = await updateLastActivity(newHabit.category_name);
        console.log("category_name: ", newHabit.category_name);
        console.log("Last activity update: ", lastActivityUpdate);
        if (!lastActivityUpdate) {
            console.error("Failed to update last activity");
            // Optionally handle this failure depending on your application needs
        }

        // Update points based on the new habit

        const { points, coins } = calculatePoints(newHabit);

        updatePointswithChange(newHabit.category_name, {
            points: points,
            coins: coins,
        });

        return habit_response.data;
    } catch (err) {
        if (err.response && err.response.status === StatusCodes.UNAUTHORIZED) {
            logout(); // Session is expired/invalid, so logout
        }
        return null;
    }
}

/**
 * Calculates the points and coins for a new habit.
 * @param {JSON} newHabit - The new habit to calculate points for.
 * @returns {JSON} - The points and coins for the new habit.
 */
export function calculatePoints(newHabit) {
    let points = 0;
    let coins = 0;

    switch (newHabit.category_name) {
        case "Exercising":
            // 5 points per 5 minutes (for only full 5 minutes)
            points +=
                Math.floor(newHabit.details.workout.workout_duration / 5) * 5;

            // default 3 coins
            coins += 9;

            // 1 additional coin for medium intensity, high intensity, and over 30 minutes
            if (newHabit.details.workout.workout_intensity == "Medium") {
                coins += 3;
            }
            if (newHabit.details.workout.workout_intensity == "High") {
                coins += 3;
            }
            if (newHabit.details.workout.workout_duration > 30) {
                coins += 3;
            }

            break;

        case "Eating":
            // default points
            points += 9;

            if (newHabit.details.eating.eating_tag == "Snack") {
                coins += 6;
            } else {
                coins += 12;
            }

            if (newHabit.details.eating.healthy_meal) {
                coins += 9;
            }

            break;

        case "Studying":

            // 4 points per 30 minutes (for only full 5 minutes)
            points +=
                Math.floor(newHabit.details.study.study_duration / 30) * 4;

            // default coins
            coins += 6;

            if (newHabit.details.study.study_duration > 30) {
                coins += 6;
            }

            if (newHabit.details.study.study_duration > 60) {
                coins += 3;
            }

            if (newHabit.details.study.study_productivity == "Medium") {
                coins += 6;
            }

            if (newHabit.details.study.study_productivity == "High") {
                coins += 6;
            }



            break;

        case "Sleeping":
            // 4 points per hour (for only full hour)
            points +=
                Math.floor(newHabit.details.sleep.sleep_duration) * 4;

            if (newHabit.details.sleep.is_nap) {
                coins += 6;
            } else {
                coins += Math.floor(newHabit.details.sleep.sleep_duration) * 3; // 3 coins per hour
            }

            break;
    }

    return { points, coins };
}

/**
 * Retrieves habits by category for the logged-in user.
 * @param {String} category_name - The category name to retrieve habits for.
 * @returns {Array|null} - An array of habit objects if successful; otherwise, null.
 */
export async function retrieveHabitsByCategory(category_name) {
    try {
        const response = await Api.get(`/habits/${category_name}`);

        if (
            response.status === StatusCodes.OK &&
            response.data &&
            response.data.habits
        ) {
            return response.data.habits; // Return the habits if the call was successful
        }

        // If the response status is OK but no data found, return null
        return null;
    } catch (err) {
        // Handle specific HTTP errors
        if (err.response) {
            if (err.response.status === StatusCodes.UNAUTHORIZED) {
                logout(); // Session is expired/invalid, so logout
                return null; // Return after logout to stop further execution
            }

            if (err.response.status === StatusCodes.NOT_FOUND) {
                return []; // Return an empty array if no habits are found
            }
        }

        console.error("Error retrieving habits by category:", err);
        return null;
    }
}

/**
 * Deletes a habit for the user.
 * @param {String} habitId - The ID of the habit to delete.
 * @returns {boolean} - True if the habit was successfully deleted; otherwise, false.
 */
export async function deleteHabit(habitId) {
    try {
        const response = await Api.delete(`/habits/${habitId}`);

        if (response.status === StatusCodes.OK) {
            return true; // Habit was successfully deleted
        }

        // If the response status is OK but action was not successful, log the issue and return false
        console.error("Error deleting habit:", response.data);
        return false;
    } catch (err) {
        if (err.response && err.response.status === StatusCodes.UNAUTHORIZED) {
            logout(); // Session is expired/invalid, so logout
        } else {
            console.error("Error deleting habit:", err);
        }
        return false;
    }
}

/**
 * Retrieves the total number of habits for all categories for the logged-in user.
 * @returns {number} - The total count of habits across all categories.
 */
export async function getTotalHabits() {
    const categories = ["Exercising", "Sleeping", "Eating", "Studying"];
    let totalHabits = 0;

    // Retrieve habits for each category and accumulate the counts
    for (const category of categories) {
        const habits = await retrieveHabitsByCategory(category);
        if (habits) {
            totalHabits += habits.length;
        }
    }

    return totalHabits;
}
