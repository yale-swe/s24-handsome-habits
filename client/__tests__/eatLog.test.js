import React from "react";
import { fireEvent, render, act } from "@testing-library/react-native";
import EatLog from "../src/screens/eatLog";
import * as habitService from "../src/services/habitService";
import AsyncStorage from "@react-native-async-storage/async-storage";

// Mocks
jest.mock("../src/services/habitService", () => ({
    addHabit: jest.fn(),
}));

jest.mock("@react-native-async-storage/async-storage", () => ({
    getItem: jest.fn(),
}));

jest.mock("axios", () => ({
    defaults: {
        headers: { Cookie: "" },
    },
}));

describe("EatLog Page", () => {
    const mockNavigate = jest.fn();

    const eatSetup = () => {
        const utils = render(
            <EatLog navigation={{ navigate: mockNavigate }} />
        );
        return {
            ...utils,
        };
    };

    beforeEach(() => {
        // Clear all instances and calls to constructor and all methods:
        habitService.addHabit.mockClear();
        AsyncStorage.getItem.mockClear();
        mockNavigate.mockClear();
        jest.useFakeTimers("modern");
        jest.setSystemTime(new Date(2020, 3, 1, 9, 30));
    });

    // Test if all components are rendered correctly
    it("renders all components correctly", () => {
        const { getByText, getByTestId } = eatSetup();

        // Check for all labels
        expect(getByText("Healthy?")).toBeTruthy();
        expect(getByText("Meal Type")).toBeTruthy();
        expect(getByText("Time")).toBeTruthy();
        expect(getByText("Add Meal")).toBeTruthy();

        // Check for all components
        expect(getByTestId("BackButton")).toBeTruthy();
        expect(getByTestId("TitleInput")).toBeTruthy();
        expect(getByTestId("TimeSelect")).toBeTruthy();
        expect(getByTestId("HorizontalSelect")).toBeTruthy();
        expect(getByTestId("DescriptionInput")).toBeTruthy();
        expect(getByTestId("AddHabitButton")).toBeTruthy();
    });

    // Test if the back button navigates back to the Eat screen
    it("navigates back to Eat screen on back button press", () => {
        const { getByTestId } = eatSetup();

        const backButton = getByTestId("touchableBack");
        fireEvent.press(backButton); // click the button part of component
        expect(mockNavigate).toHaveBeenCalledWith("Eat"); // check if navigate was called with correct argument
    });

    // Test if the logEat function works correctly
    it("correctly logs eat and navigates", async () => {
        const { getByTestId } = eatSetup();

        // Assume you trigger logEat via some UI interaction, e.g., pressing a "Log Eat" button
        fireEvent.press(getByTestId("AddHabitButton"));

        await act(async () => {
            expect(AsyncStorage.getItem).toHaveBeenCalledWith("cookies"); // check async storage was called

            // Check addHabit was called with the expected newEat object
            expect(habitService.addHabit).toHaveBeenCalledWith(
                expect.objectContaining({
                    title: expect.any(String),
                    category_name: "Eating",
                })
            );
        });

        // Verify navigation
        expect(mockNavigate).toHaveBeenCalledWith("Eat");
    });

    it("matches the snapshot", () => {
        const { toJSON } = eatSetup();
        expect(toJSON()).toMatchSnapshot();
    });
});
