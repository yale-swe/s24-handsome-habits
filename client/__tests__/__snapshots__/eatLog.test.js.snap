// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EatLog Page matches the snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FBCD5A",
      "flex": 1,
      "paddingTop": 65,
    }
  }
>
  <View
    style={
      {
        "alignSelf": "flex-start",
        "marginBottom": 15,
        "marginLeft": 15,
        "marginTop": 10,
      }
    }
  >
    <View
      testID="BackButton"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#0D3A4F",
            "borderRadius": 5,
            "color": "white",
            "height": 29,
            "justifyContent": "center",
            "margin": 10,
            "opacity": 1,
            "width": 28,
          }
        }
        testID="touchableBack"
      >
        <Image
          source={
            {
              "testUri": "../../../src/assets/images/backarrow.png",
            }
          }
          style={
            {
              "height": 18,
              "margin": 3,
              "width": 18,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 10,
        "height": "80%",
        "padding": 20,
        "paddingBottom": 30,
        "paddingTop": 30,
        "width": "85%",
      }
    }
  >
    <View
      style={
        {
          "marginBottom": 20,
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Title"
        style={
          {
            "backgroundColor": "#FEF5DE",
            "borderRadius": 5,
            "fontFamily": "OpenSans-Bold",
            "fontSize": 20,
            "height": 50,
            "padding": 10,
            "textAlign": "left",
          }
        }
        testID="TitleInput"
        value=""
      />
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "marginBottom": 20,
        }
      }
    >
      <Text
        style={
          {
            "fontFamily": "OpenSans-Bold",
            "fontSize": 16,
            "textAlign": "left",
          }
        }
      >
        Time
      </Text>
      <View
        testID="TimeSelect"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#cce7f3",
              "borderRadius": 5,
              "height": 25,
              "justifyContent": "center",
              "marginStart": 10,
              "opacity": 1,
              "paddingEnd": 10,
              "paddingStart": 10,
              "textAlign": "center",
            }
          }
          testID="touchableTime"
        >
          <Text
            style={
              {
                "color": "grey",
              }
            }
          >
            9:30am
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "marginBottom": 20,
        }
      }
    >
      <Text
        style={
          {
            "fontFamily": "OpenSans-Bold",
            "fontSize": 16,
            "textAlign": "left",
          }
        }
      >
        Meal Type
      </Text>
      <RCTScrollView
        contentContainerStyle={
          {
            "flexDirection": "row",
            "paddingLeft": 0,
            "paddingRight": 10,
            "paddingTop": 10,
          }
        }
        horizontal={true}
        showsHorizontalScrollIndicator={false}
        testID="HorizontalSelect"
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#cce7f3",
                "borderColor": "#8ECAE6",
                "borderRadius": 30,
                "borderStyle": "dashed",
                "borderWidth": 2,
                "margin": 5,
                "opacity": 1,
                "padding": 5,
                "paddingHorizontal": 20,
              }
            }
            testID="touchableHorizontal"
          >
            <Text>
              Snack
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#cce7f3",
                "borderColor": "#8ECAE6",
                "borderRadius": 30,
                "borderStyle": "dashed",
                "borderWidth": 2,
                "margin": 5,
                "opacity": 1,
                "padding": 5,
                "paddingHorizontal": 20,
              }
            }
            testID="touchableHorizontal"
          >
            <Text>
              Breakfast
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#cce7f3",
                "borderColor": "#8ECAE6",
                "borderRadius": 30,
                "borderStyle": "dashed",
                "borderWidth": 2,
                "margin": 5,
                "opacity": 1,
                "padding": 5,
                "paddingHorizontal": 20,
              }
            }
            testID="touchableHorizontal"
          >
            <Text>
              Lunch
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#cce7f3",
                "borderColor": "#8ECAE6",
                "borderRadius": 30,
                "borderStyle": "dashed",
                "borderWidth": 2,
                "margin": 5,
                "opacity": 1,
                "padding": 5,
                "paddingHorizontal": 20,
              }
            }
            testID="touchableHorizontal"
          >
            <Text>
              Dinner
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#cce7f3",
                "borderColor": "#8ECAE6",
                "borderRadius": 30,
                "borderStyle": "dashed",
                "borderWidth": 2,
                "margin": 5,
                "opacity": 1,
                "padding": 5,
                "paddingHorizontal": 20,
              }
            }
            testID="touchableHorizontal"
          >
            <Text>
              Other
            </Text>
          </View>
        </View>
      </RCTScrollView>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "marginBottom": 20,
        }
      }
    >
      <Text
        style={
          {
            "fontFamily": "OpenSans-Bold",
            "fontSize": 16,
            "textAlign": "left",
          }
        }
      >
        Healthy?
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="touchableToggle"
      >
        <View
          style={
            [
              {
                "borderRadius": 25,
                "height": 25,
                "justifyContent": "center",
                "marginStart": 10,
                "width": 50,
              },
              {
                "alignItems": "flex-start",
                "backgroundColor": "#D6D6D6",
              },
            ]
          }
        >
          <View
            style={
              {
                "backgroundColor": "white",
                "borderRadius": 10,
                "height": 18,
                "margin": 4,
                "width": 18,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        {
          "marginBottom": 0,
        }
      }
    >
      <TextInput
        multiline={true}
        onChangeText={[Function]}
        placeholder="Description"
        style={
          [
            {
              "backgroundColor": "#FEF5DE",
              "borderRadius": 5,
              "height": 125,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
              "textAlign": "left",
              "textAlignVertical": "top",
            },
          ]
        }
        testID="DescriptionInput"
        value=""
      />
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "bottom": 30,
          "justifyContent": "flex-end",
          "left": 0,
          "marginTop": 20,
          "position": "absolute",
          "right": 0,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#0D3A4F",
            "borderRadius": 10,
            "color": "white",
            "height": 50,
            "justifyContent": "center",
            "opacity": 1,
            "padding": 10,
            "width": 250,
          }
        }
        testID="AddHabitButton"
      >
        <Text
          style={
            {
              "color": "white",
              "fontFamily": "OpenSans-Bold",
              "fontSize": 20,
              "textAlign": "center",
            }
          }
        >
          Add Meal
        </Text>
      </View>
    </View>
  </View>
</View>
`;
